gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java

```diff
@@ -160,16 +181,32 @@ protected JobState getJobState()
   }
 
   public void cancel() {
-    if (_taskAttempt != null) {
+    int retryCount = 0 ;
+    int maxRetry = this._dynamicConfig.hasPath(MAX_RETRY_WAITING_FOR_INIT_KEY)
+        ? this._dynamicConfig.getInt(MAX_RETRY_WAITING_FOR_INIT_KEY) : DEFAULT_MAX_RETRY_WAITING_FOR_INIT;
+
+    try {
+      _lock.lock();
       try {
+        while (_taskAttempt == null) {
+          // await return false if timeout on this around
+          if (!_taskAttemptBuilt.await(5, TimeUnit.SECONDS) && ++retryCount > maxRetry) {
+            throw new IllegalStateException("Failed to initialize taskAttempt object before cancel");
```

Comment:
Will this fail the job? Or we just want to log the error and ignore.

Reply:
It will fail the cancel method and propagate exception back to Helix. 

The reason of doing so, is because it is potentially dangerous to proceed in this case where the cancel was issue but the `taskAttempt` has not been initialized. For example it could result in two task running in the same helix instance in our streaming-ingestion pipeline where both task not falling behind due to the lack of enough resources